runtime/parachains/src/session_info.rs

```diff
@@ -0,0 +1,256 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! The session info module provides information about validator sets
+//! from prior sessions needed for approvals and disputes.
+//!
+//! See https://w3f.github.io/parachain-implementers-guide/runtime/session_info.html.
+
+use primitives::v1::{AuthorityDiscoveryId, SessionIndex, SessionInfo};
+use frame_support::{
+	decl_storage, decl_module, decl_error,
+	weights::Weight,
+};
+use crate::{configuration, paras, scheduler};
+use sp_std::{cmp, vec::Vec};
+
+pub trait Trait:
+	frame_system::Trait
+	+ configuration::Trait
+	+ paras::Trait
+	+ scheduler::Trait
+	+ AuthorityDiscoveryTrait
+{
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as ParaSessionInfo {
+		/// The earliest session for which previous session info is stored.
+		EarliestStoredSession get(fn earliest_stored_session): SessionIndex;
+		/// Session information in a rolling window.
+		/// Should have an entry in range `EarliestStoredSession..=CurrentSessionIndex`.
+		/// Does not have any entries before the session index in the first session change notification.
+		Sessions get(fn session_info): map hasher(identity) SessionIndex => Option<SessionInfo>;
+	}
+}
+
+decl_error! {
+	pub enum Error for Module<T: Trait> { }
+}
+
+decl_module! {
+	/// The session info module.
+	pub struct Module<T: Trait> for enum Call where origin: <T as frame_system::Trait>::Origin {
+		type Error = Error<T>;
+	}
+}
+
+/// An abstraction for the authority discovery pallet
+/// to help with mock testing.
+pub trait AuthorityDiscoveryTrait {
+	/// Retrieve authority identifiers of the current and next authority set.
+	fn authorities() -> Vec<AuthorityDiscoveryId>;
+}
+
+impl<T: pallet_authority_discovery::Trait> AuthorityDiscoveryTrait for T {
+	fn authorities() -> Vec<AuthorityDiscoveryId> {
+		<pallet_authority_discovery::Module<T>>::authorities()
+	}
+}
+
+impl<T: Trait> Module<T> {
+	/// Handle an incoming session change.
+	pub(crate) fn initializer_on_new_session(
+		notification: &crate::initializer::SessionChangeNotification<T::BlockNumber>
+	) {
+		let config = <configuration::Module<T>>::config();
+
+		let dispute_period = config.dispute_period;
+		let n_parachains = <paras::Module<T>>::parachains().len() as u32;
+
+		let validators = notification.validators.clone();
+		let discovery_keys = <T as AuthorityDiscoveryTrait>::authorities();
+		// FIXME: once we store these keys: https://github.com/paritytech/polkadot/issues/1975
```

Comment:
please correct me if I'm wrong here, but my assumption is that approval voting subsystem will store `ApprovalId`s via `OneSessionHandler` impl